Added the fixes

Taure opened the pull request:
This is your branch and mine. Hope it don't break anything for you.

kovyl2404 committed:
Fix windows-related issues

kovyl2404 committed:
Fix windows-related issues

kovyl2404 committed:
Fix dialyzer tests on windows

kovyl2404 committed:
Fix dialyzer tests on windows

tsloughter committed:
improve escriptize provider

ferd committed:
Merge pull request #501 from tsloughter/escriptize_improvements

improve escriptize provider

Taure committed:
Added code from my branch.

Fix issues with copies and some symlink issues, also add touch.

Taure committed:
Merge branch 'kovyl2404' of https://github.com/Taure/rebar3 into kovyl2404

Conflicts:
	src/rebar_file_utils.erl

Taure commented:
Hope this fix the issues.

kovyl2404 commented on src/rebar_utils.erl:
```diff
@@ -158,11 +158,21 @@ sh(Command0, Options0) ->
-        {error, {_Rc, _Output}=Err} ->
-            ErrorHandler(Command, Err)
+    case os:type() of
+        {win32, _} ->
```
So again, why do you need this? Isn't it enough to expect that `rebar_utils:sh("robocopy blah blah"...)` may return something different from `{ok, _}`?

Taure replied:
Different is on os.

*nix system when doing copy returns 0 if it is ok. Robocopy returns values from 0 - 8 and 16 if nothing was copies (This can be an ok or sometimes considered a error)

so using the sh_loop will fail when doing copy. I thought it would be better with a own sh_loop for windows handling exit codes than in functions using robocopy do.

Res = rebar_utils:sh("robocopy blah blah"...) 
case Res of
 {error, {1, []}} -> ...

I thought it would be better to handle the codes that is ok and fail if we get something else.

If Robocopy copies files into a directory where there is already files in it it can give a exit code 3. and this will fail if only is looking for a 1.

That is why I did this. Not exactly sure if it is needed. :)

Also I followed how things was working for unix systems.

kovyl2404 replied:
However, `rebar_utils:sh` could be used to run another things, not only robocopy. And I believe in this case such behaviour could be harmful. I think it is better to handle all possible robocopy exit codes on higher level, in code which calls `rebar_utils:sh`.

@ferd, sorry for disturbance. What do you think about?

Taure replied:
Hmm.. true.

I see that it uses it for git, hg and so..

Need to go to bed now. But if you want you can remove the sh_loop_win32 and add the checks in xcopy_win32 instead and the mv function?

kovyl2404 commented:
Sorry, There is midnight in my TZ.  I'll continue tomorrow.

Taure committed:
Moved Robocopy logic to xcopy

Taure commented:
I deleted the sh_loop_win32 and moved that logic to xcopy_win32.

After some research git uses 0 for ok and 1 as error and this would have disturbed outcome on windows machines.

kovyl2404 commented on src/rebar_file_utils.erl:
```diff
@@ -198,17 +218,18 @@ delete_each_dir_win32([Dir | Rest]) ->
-        false ->
+    %% "xcopy \"~s\" \"~s\" /q /y /e 2> nul", Chanegd to robocopy to
+    %% handle long names. May have issues with older windows.
+    {_, {Rc, _}} = rebar_utils:sh(
```
Can it return `{ok, []}` as well?

Taure replied:
Maybe move it to the case below? For robocopy? 

So it does something like:
Case res of
 {_,{rc, -}} when rc < 9; rc =:= 16 ->... 
{ok, []} -> ok... 

I'm at work and writing on a mobile. :)

kovyl2404 replied:
I've added `rebar_file_utils:win32_robocopy_ok/1` to my branch to do this stuff.

kovyl2404 commented:
Daniel, frankly I've completely confused with all those commits, merges and so on. I've rebased old your [commit](da598d7614102c2da8f0657ed338f94a6872aeb3) to my branch head. Could you review it, find which your changes I've missed and commit it again - just one commit to the branch head. Without any complex merges and so on. I think it would be much easier for rebar3 guys to track changes and review this changes.

Taure commented:
Sure. Will fix it after work. Sorry for the mess. :)

Taure commented:
I close this one and create a new one from head of branch. With fixes.

kovyl2404 commented:
@Taure, I've rewrite history in my branch. Added some fixes and squashed it with another. So be careful. I think it is better to completely to delete _my_ branch and then fetch and checkout it again.

Taure commented:
That is the plan. When I come home I will check out a fresh branch and then fix my part in one commit. Then do a pull request.

Taure commented:
But will do it later tonight.
